io/common/include/detray/io/file_handle.hpp

```diff
@@ -0,0 +1,66 @@
+/** Detray library, part of the ACTS project (R&D line)
+ *
+ * (c) 2023 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+// System include(s)
+#include <cassert>
+#include <cstdint>
+#include <fstream>
+#include <ios>
+#include <iostream>
+#include <limits>
+#include <string>
+
+namespace detray::io::detail {
+
+/// Wrapper around a file stream
+class file_handle final {
+
+    public:
+    /// All writers must define a file name
+    file_handle() = delete;
+
+    /// File gets created with a @param name and @param extension
+    file_handle(const std::string &name = "",
+                const std::string &extension = "txt",
+                std::ios_base::openmode mode = std::ios_base::in |
+                                               std::ios_base::out) {
+        // Default name
+        std::string file_name{name.empty() ? "detray_" + std::to_string(n_files)
+                                           : name};
+
+        // Count the new file
+        ++n_files;
+        assert(n_files < std::numeric_limits<std::uint16_t>::max());
+
+        // Open file
+        m_stream.open(file_name + "." + extension, mode);
+
+        if (!m_stream.is_open()) {
+            std::cout << "Could not open file " << file_name + "." + extension
+                      << std::endl;
+        }
+    }
+
+    /// Destructor closes the file
+    ~file_handle() { m_stream.close(); }
+
+    /// @returns the output stream
+    std::fstream &operator*() { return m_stream; }
+
+    private:
+    /// Output file handle
+    std::fstream m_stream;
+
+    /// How many files have been created? Maximum: 65'536
+    static std::uint16_t n_files;
```

Comment:
What's the exact point of this counter? :confused: You don't decrement it in the destructor... so what's it supposed to keep track of exactly...?